Revamp the patterns proposal.

munificent opened the pull request:
There are a large number of changes here, but structural and meaningful.

- Unify binder and matcher patterns into a single grammar.
- Always treat simple identifiers as variables in patterns.
- Change the `if (expr case pattern)` syntax.
- Change the guard syntax.
- Record patterns match only record objects.
- New patterns for relational operators, `|`, `&`, and `(...)`.
- Get rid of explicit wildcard patterns since they're redundant with
  untyped variable patterns named `_`.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-```
-mapBinder ::= mapTypeArguments? '{' mapBinderEntries '}'
+**TODO: Allow a `...` element in order to ignore some positional fields while
+capturing the suffix.**
````
Probably not!

Why only ignore positional fields? It should at least also allow ignoring named fields.

I'd me much more at ease about ignoring only *trailing* positional fields.
Otherwise it becomes arbitrary that you can skip some fields, but only once.
```dart
  if (var (x, ..., y) = someRecord) ... // OK
  if (var (x, ..., y, ..., z) = someRecord) // Not OK?
```
You can only have one trailing position, but you can have multiple ranges in the middle of a record.


It means the pattern now abstracts over the width of the record. A pattern like `(x, ..., z)` can match both `(1, 2)` an `(1, 2, 3)`. That means that we need a run-time representation of the structure and a way to abstract over it and access members, without knowing the structure.

If we don't do this, and only allow field access on known record structures, we should be able to keep runtime structure information to a minimum.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+log(Severity severity, String message) {
+  switch (severity) {
+    case Severity.error(prefix: p):
+      print('!! $p !! $message'.toUppercase());
```
That seems a little weird. The `Severity.error.prefix` is a fixed constant.
Why do you want to match the enum, and still bind the prefix, instead of just:
```dart
  case Severity.error: 
      print('!! ${Severity.error.prefix} !! $message'.toUpperCase());
``` 
or, if you really want a shorter way to refer to it:
```dart
  case Severity.error & e: 
      print('!! ${e.prefix} !! $message'.toUpperCase());
```
(The `& e` is one way to get a short name for anything you have matched, even those patterns which don't allow binding themselves).

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-mapBinderEntries ::= mapBinderEntry ( ',' mapBinderEntry )* ','?
-mapBinderEntry ::= expression ':' binder
+```
+extractorPattern ::= extractorName typeArguments? '(' patternFields? ')'
````
If you want to be able to match enum values, not just types, then you need to be able to match `ClassName<typeArgs>.name`.

So maybe `extractorName typeArguments? ('.' id)?` instead, to account for the enum element name?

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-a named field. In that case, the field implicitly contains a variable binder
-subpattern with the same name. These are equivalent:
+You can also use an extractor to both match an enum value and destructure
+fields from it:
```
Doesn't seem so from the grammar above, so is this really intended?

It seems to be mixing the constant pattern and the extractor pattern.

Can other constant patterns use it as well, or is it only `enum` declared constants?
(So not user-defined enum-like classes?)

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-These often occur as subpatterns of a destructuring pattern in order to capture
-a destructured value.
+*   A type argument list is present and does not match the arity of the type of
+    `extractorName` or `extractorName` is an enum value.
```
Ah, so enums patterns cannot have type arguments, but these days enum values can.
Seems wrong that you can't do `case MyEnum<final T>.thing(:T defaultValue)`.

(Still feels odd to destructure an enum to begin with, but at least allow it completely.)

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+**TODO: Specify this restriction more precisely.**
+
+**TODO: Should we expand this to allow type patterns in for-loop variables,
+fields, or top-level variables?**
```
Local declarations in general should be allowed to be patterns. That includes loop variables of a `for`/`in` and *parameters declarations*.
So, definitely allow 
```dart
for (List<final T> list in listOfLists as List<List<Object?>>) ...
```
and 
```dart
Set<S> toSet<S>(List<final T extends S> list) => <T>{...list};
```

I'd not allow patterns for "non-scoped" declarations (top-level, static, instance variables), because code accessing those are not necessarily dominated by the variable initialization. Pattern matching and binding is a scope based behavior, it only makes sense when there is a control flow that it can follow.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+var {1: a} = {1: 2};                 // Map.
+var (a, b, x: x) = (1, 2, x: 3);     // Record.
+var Point(x: x, y: y) = Point(1, 2); // Extractor.
+var string? = nullableString;        // Null-check.
```
This one would not be allowed, because the pattern is refutable. The pattern doesn't match `null`, so `var string? = null;` would not bind `string`, which is not OK for a declaration.

Did you mean `var string! = ...`?

munificent replied:
Good catch. Wrapped in a pattern-if to make it valid.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+await switch (n) {
+  case 1 => aFuture;
+  case 2 => anotherFuture;
+};
```
These expression switches are invalid because they are not exhaustive.

munificent replied:
Fixed.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-fields from it:
+Over half of the switch cases in a large corpus of packages contain either a
+single return statement or an assignment followed by a break so there is some
+evidence this will be useful.
```
Let's consider, once more, whether we want `=>` instead of `:` here.

It definitely makes the expression switch stand out as different from the statement switch.
Just want to think a little harder about whether that's a good thing or a bad thing.

Switching from one to the other requires changing every `:` to `=>` ... but it's probably `: return`, so ... maybe it makes sense.
But `=>` usually means `return`, and here it doesn't.
... conflicted...!

munificent replied:
Yeah, I'm on the fence. There's an open issue discussing this, so I think we can leave it as-is in this PR.

lrhn reviewed (commented):
Only got to 1124 today.

Still looks awesome (with contentious details :grin: ).

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1682,130 +1887,138 @@ At runtime, a pattern is matched against a value. This determines whether or not
-            syntax that that infers an implicit variable subpattern from the
-            field's name, match `r` against that inferred variable subpattern.
+*   **Literal** or **constant**: The pattern matches if `o == v` evaluates to
+    `true` where `o` is the pattern's value.
```
It's probably a good call on using `o == v`. The pattern decides whether the value matches, not the other way around.

Using `==` is still dangerous, since it calls user code and can act erratically.
It does mean that `c` is the same as `== c`, which is convenient.
It does mean there is no easy way to check for identity.
Guess you can always do:
```dart
class Ider {
  final Object? value;
  Ider(this.value);
  int get hashCode => identityHasHCode(value);
  bool operator==(Object other) => identical(value, other);
}
```
and use that in your patterns. 
(Which is why `v == o` might also be a valid choice, because then the pattern author can fiddle with equality.)

munificent replied:
Yeah, I think in practice this will almost always be used with types that have entirely boring equality behavior, so it doesn't matter too much what we do, as long as it's specified precisely.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1682,130 +1887,138 @@ At runtime, a pattern is matched against a value. This determines whether or not
+
+*   **Record**:
+
+    1.  If `v` is not a record with the same type as `p`, then the match fails.
```
Probably "same shape", not "type".

munificent replied:
Ah, good catch. In the initial version of this, yes, the record type was too precise and could fail unnecessarily.

I've tweaked the rules for calculating the record pattern's static type to mirror the matched value's type, so I think it's safe to use "same type" here. (And it avoids needing to define some runtime notion of checking a record's shape.)

The only failure mode is when the matched value isn't a record type. For that, when calculating the record pattern's type, *there* I say that we synthesize a record type with the same shape as the pattern and `Object?` for all of the fields. That way, the record pattern itself only checks the shape, and then the field subpatterns take it from there.

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -1836,8 +2049,134 @@ main() {
+```dart
+main() {
+  late var (a, b) = (1, 2);
+  a = 3;
````
Well, that's awkward!

Consider *very strongly* not to allow late patterns! Just to avoid this!

munificent replied:
Agreed!

lrhn commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1836,8 +2049,134 @@ main() {
+    *   And patterns
+    *   Grouping patterns
+
+*   **Type patterns.** This is a mostly orthogonal feature, though a useful one
```
Also the title in line 2121, covering something else.

munificent replied:
Done.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+  List<final E> _ = items;
+  return <E>{...items};
+}
+```
````
I would like to eventually support patterns in parameter lists, but given how insane the parameter list grammar already is and how big the patterns feature already is, I think we should defer it to a future release.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+| [Cast][castPattern] | `foo as String` |
+| [Grouping][groupingPattern] | `(subpattern)` |
+| [List][listPattern] | `[subpattern1, subpattern2]` |
+| [Map][mapPattern] | `{"key": subpattern}` |
```
Done.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-with a `var` or `final` keyword. Also, a variable declaration using a pattern
-can only have a single declaration "section". No comma-separated multiple
-declarations like:
+An or pattern does not have to appear at the top level of a pattern. It can be
```
Done. Did "logical-or" and "logical-and" since other patterns don't make "pattern" part of the name.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
 ```dart
-var a = 1, b = 2;
+// Matches a two-element list whose first element is 'a' or an int:
+if (var ['a' | int _, c] = list) ...
````
Yeah, in practice, I expect nested `|` patterns to be fairly rare.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+    other branch. These variable patterns can appear as subpatterns anywhere in
+    each branch, but in total both branches must contain the same variables with
+    the same types. This way, variables used inside the body covered by the
+    pattern will always be initialized to a known type.
```
I don't know, that feels pretty dubious to me. If users write a type annotation on a variable, I think they expect the variable to end up with that type. It also seems unlikely that they'd want to test a more specific type than the body of the case actually needs.

If they really did want that, a marginally better workaround is:

```dart
case BoundingShape(shape: var shape & Box()) | BoundingShape(shape: var shape & Circle()): ... 
```

In general, I feel like we usually end up regretting using LUB. I don't want bug reports from users where they're like, "Why does this variable in the pattern have type 'EfficientLength' when I'm declaring it as a List?" and we have to point out that somewhere in the long list of `|` clauses they have one variable whose type is just different enough to confuse the LUB.

Let's keep it simple and strict for now and we can always loosen it to LUB in the future if it turns out that users want it.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+using any of the equality or relational operators: `==`, `!=`, `<`, `>`, `<=`,
+and `>=`. The pattern matches when calling the appropriate operator on the
+matched value with the constant as an argument returns `true`. It is a
+compile-time error if `relationalExpression` is not a valid constant expression.
```
I like that. I updated the runtime semantics section with this your first suggestion and the static type section with your second suggestion. Thanks!

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+In Dart today, this prints "no match". With this proposal, it changes to
+"match". However, looking at the 22,943 switch cases in 1,000 pub packages
+(10,599,303 lines in 34,917 files), I found zero case expressions that were
+collection literals.
```
Haha, I found a corner of the language you didn't know! \o/

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+**Breaking change:** This is a breaking change for simple identifiers that
+appear in existing switch cases. Fortunately, it turns out that most switch
+cases are not simple named constants. I analyzed a large corpus of Pub packages
+and Flutter apps (13M+ lines in 61,346 files):
```
Yeah, it's not tiny, but if I recall, the majority of those 2,855 cases all come from uses of the charcode package. If we tweak that package such that it's imported with a prefix, the number of cases is much smaller.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+
+```dart
+if ((int x, String s) = record) {
+  print('First field is int $x and second is String $s.');
````
That's not in this update, but I have a TODO to look into that syntax. I think it's straightforward, but I didn't want to bundle too much into this giant PR.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-And turns it into:
+A literal pattern determines if the value is equivalent to the given literal
+value. There are no list and map *literal* patterns since there are actual list
+and map patterns.
```
Good call. I think, yes, literal and constant patterns should have a context type. Updated.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+  const localConstant = 1;
+  switch (2) {
+    case == localConstant: print('matched!');
+    default: print('unmatched');
```
They do the same thing, but you have to use `==` to compare to a constant with a bare identifier.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
 
-#### Null-assert binder
+**TODO: Do we want to support function types? If so, how do we handle
+first-class generic function types?**
```
Ooh, lots of interesting dragons to slay here.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-Swift][swift null check]:
+```dart
+if (var [int x, int y] = json) return Point(x, y);
+```
````
`json is var [int x, int y]` is parseable. The problem is that `json is int` is also parseable, but means something very different from `case int`. The former is a type test and the latter is a variable binder.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+
+ifCondition ::= expression // Existing if statement condition.
+              | patternDeclaration
+              | type identifier '=' expression
```
Yeah, I always intended to support that. Added a TODO. It should be straightforward, but I don't want to bother until the more fundamental stuff is sorted out.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+
+ifCondition ::= expression // Existing if statement condition.
+              | patternDeclaration
+              | type identifier '=' expression
```
Yes, exactly. Not super useful with an irrefutable pattern like your example here, but definitely useful when the type annotation is a subtype:

```dart
class C {
  Object field;

  void method() {
    if (int i = field) {
      // `i` is int here...
    }
  }
}
```

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
-    subpatterns. If a named field uses the shorthand syntax to infer a variable
-    subpattern with the same name as the field, then the type schema is `?` for
-    that field.
+*   **And**: The greatest lower bound of the context type schemas of the branches.
```
Yeah, I think you'd get `Never`. Basically, I defer to what our existing spec for GLB says here. :)

I don't think having `Never` appear here is problematic, but I'm definitely not sure. It won't do much as far as type inference goes, but that's OK. It could still be reasonable to allow it, for example:

```dart
class A {}
class B {}

class C implements A, B {}

main(Object obj) {
  if (var (A a & B b) = obj) {
    // Can use a and b here...
  }
}
```

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
+        `Object?`. *We ignore the type annotation in a refutable context because
+        the type is being used to test the value's type and shouldn't also
+        influence it. This is consistent `is` expressions where we don't use the
+        type annotation as a context type when inferring its operand.*
```
Heh, yes, but I'm not trying to fix all the mistakes. :)

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
+    *   If `p` has a type argument, then `E` is the type argument.
+
+    *   Else `E` is the greatest lower bound of the type schemas of all element
+        subpatterns. *We use the greatest lower bound to ensure that the outer
```
Now that I look into it, I don't think we define LUB and GLB of type schemas, as far as I can tell. :-/

@leafpetersen, is it OK to use GLB here?

munificent replied:
For now, I'll add a TODO to the proposal.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
+*   **Relational**:
+
+    1.  It the operator is a comparison (`<`, `<=`, `>`, or `>=`), then it is a
+        compile-time error if `M` does not define that operator, or if the type
```
Done.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-  | extractMatcher
-  | nullCheckMatcher
+```dart
+var [a] = [1], (b, c) = (2, 3);
````
Good idea. Done.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+```
+topLevelDeclaration ::=
+  | // Existing productions...
+  | 'late'? patternDeclaration ';' // New.
````
I think we could make this work, but I agree it gets pretty weird. For now, I've gone ahead and removed support for late pattern variables. This also means removing support for patterns in top-level variables and static fields since those are also lazy-initialized. I think that's OK. We can always expand to support that in a future release.

munificent replied:
Removed support for late.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1197,7 +1317,7 @@ switch case, and if-case statements, there is no downwards inference from
 value is inferred with no downwards context type and we jump straight to
 inferring the types of the case patterns from that context type. *The intent of
-a matcher pattern is to query the type of the matched value, so it would be
+a refutable pattern is to query the type of the matched value, so it would be
```
Yeah, that's a good point. I rewrote this. We still don't provide a downwards inference context for switches, but that's because there are multiple cases and it isn't clear how we would merge those into a single context type schema. But we do provide a downwards inference context for both pattern variable declarations and pattern-if statements since in the latter there is only one pattern.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
     ```dart
-    var p = Point(x: 1, y: 2);
-    var (x: x, y: y) = p;
+    var [[int x]!] = [[]]; // Infers List<List<int>?> for the list literal.
````
I think of the context type of a pattern more like "the type it could reasonably *attempt* to match" but not necessarily the type it *will* match.

If the user took the time to write a null-check or null-assert pattern, that implies that they *do* expect `null` to sometimes flow into that pattern (and be refuted or throw). Otherwise, they would have simply left off the `!` or `?` entirely.

But, yes, context types on refutable patterns don't often come into play. (Though I guess they do now that I tweaked the proposal to use a context type for pattern-ifs.)

munificent replied:
That's a good point. You're right that we already say that we don't even calculate a context type for switch cases. So there really is no point in defining it at all for refutable patterns.

However, I just tweaked the proposal to say that we do use a context type for pattern-if statements, so now the context type schema does come into play with refutable patterns, but *only* for those. If there *is* going to be a context type at all for a refutable pattern, I think it may as well be the full one.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
-*   **Cast binder**, **variable binder**, **wildcard matcher**, or **extractor
-    matcher**: The context type schema is `Object?`.
+    *   If the pattern is in a refutable context, the context type schema is
+        `Object?`. *We ignore the type annotation in a refutable context because
```
We do now because I just tweaked it to allow context types in pattern-if. But I also removed these special cases because if we're going to use a context type there, we may as well use a precise one. That way, a pattern in a pattern-if that *could* be just an irrefutable destructuring can be if downwards inference makes it such:

```dart
List<T> makeList<T>() => [];

if ([int i] = makeList()) { ... }
```

Using `Object?` as the context type schema for `int i` would lead the `[int i]` pattern to have context type schema `List<Object>`. That would in turn lead downwards inference to infer `makeList<Object>`. Which means now the `[int i]` pattern would have to be refutable to check that the returned list is actually a `List<int>`.

But there's no need for that to be refutable at all if we use `int` as the context type schema for `int i`. So now it unconditionally uses a precise context type schema regardless of whether its in a refutable context.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
 
 The static type of a pattern `p` being matched against a value of type `M` is:
```
Thanks for pushing on this to get me to clarify my thinking.

Concretely:

The context type schema for a pattern is used:

*   To calculate the context type schema of its superpattern. For example, in `[int x]`, the context type schema `int` for the subpattern is used to calculate `List<int>` for the outer pattern.

*   To provide a downwards inference context type for the matched value expression in pattern variable declarations and pattern-if statements.

The static type for a pattern is used:

*   To calculate the static type of its superpattern. For example,
    in `[int x]`, the type `int` for the subpattern is used to calculate `List<int>` for the outer pattern.

*   To determine if a null-check, variable, list, map, record, or extractor pattern is refutable or not. All of those potentially perform a type test and are only irrefutable if the type being tested is a supertype of the value's type.

    A consequence of this is that when the pattern is in an irrefutable context, if the pattern's type isn't a supertype of the value type, this becomes a compile error because now the pattern is considered refutable and it's an error to have a refutable pattern in an irrefutable context.

*   At runtime, to determine if a null-check, variable, list, map, record, or extractor pattern matches its type test.

So, basically, in an irrefutable context, the pattern's static type determines if you get a compile error because the pattern is trying to do a "downcast" which might fail. In an refutable context, that downcast is allowed statically and then is checked at runtime and refutes the pattern if it fails.

> Maybe if we show how it's used (being used for the subpatterns' type inference confuses me since it seems I need the types of those subpatterns to do inference on the RHS, which happens before finding the static type of the pattern match.

The information flow is a little complex. It's:

1. Calculate the context type schemas of the leaf subpatterns.
1. From those, calculate context type schemas of parent patterns.
2. Use the context type schema of the outermost pattern to do downwards inference on the value type.
3. Apply the value's type to the outermost pattern.
4. Destructure the value type and send it down into the subpatterns.
5. Calculate the types of leaf subpatterns.
6. From those, calculate the parent pattern types.

For example:

```dart
[int i, var x] = [];
```

1. First we calculate the context type schemas for `int i` and `var x`, which yields `int` and `?`.
2. From those, we infer `List<int>` as the context type schema of the outer list pattern.
3. Use that in downwards inference to infer `List<int>` for the `[]` expression.
4. The list pattern destructures that element type `int` and uses that as the matched value type to calculate the static types of its element subpatterns.
5. Calculate `int` for the type element subpatterns (so now we know `var x` has type `int`).
6. Take the GLB of those to infer `List<int>` for the list pattern.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
+        the inner pattern using `N` as the matched value type. Otherwise,
+        calculate `q` using `M` as the matched value type. *A null-check or
+        null-assert pattern removes the nullability of the type it matches
+        against.*
```
Done.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
+        var [x!] = <int?>[]; // x is int.
+        ```
+
+    2.  The static type of `p` is `q?`. *The intent of `!` and `?` is only to
````
You're right. I didn't have a coherent notion of what the static type meant. It's clearer now and you're current that it's **NONNULL**(M).

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
+        value of its nullable base type, and not simply `Object?`.*
+
+*   **Literal** or **constant**: The static type of `p` is the static type of
+    the pattern's value expression.
```
No longer need a static type for these at all.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
+        and `Object` for `b`.*
+
+*   **Cast**: The static type of `p` is `Object?`. *Casts exist to check types
+    at runtime, so statically accept all types.*
```
No longer need a static type for this either.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
````diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
         ```
 
-        *both `a` and `b` use `Object` as their matched value type.*
+        *both `a` and `b` use `num` as their matched value type.*
````
You're right. Preferring to infer the list's type from the subpatterns leads you to unnecessarily precise list types that then get refuted even if the list does actually contain the elements you want.

Changed this to use the matched value's element type for the list pattern's type. I think that's what we want.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1374,173 +1580,172 @@ The static type of a pattern `p` being matched against a value of type `M` is:
-        a record type or `dynamic`. *Only records support positional field
-        destructuring.*
+It is a compile-time error if the type of an expression in a guard clause is not
+`bool` or `dynamic`.
```
Yes and yes.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1374,173 +1580,172 @@ The static type of a pattern `p` being matched against a value of type `M` is:
-    2.  Else the type of `p` is `M`. *This means that an untyped variable
-        pattern can have its type indirectly inferred from the type of a
-        superpattern:*
+* Top-level and local variable declarations.
```
I removed support for top-level variables and static fields when you convinced me to remove support for late variables. Now you've convinced me to drop support for instance fields too.

I guess this will give our future selves something to work on in later releases. :)

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1374,173 +1580,172 @@ The static type of a pattern `p` being matched against a value of type `M` is:
+* For loop variable declarations.
+* Switch statement cases.
+* A new switch expression form's cases.
+* A new pattern-if statement.
```
Yeah, I'm very interested in exploring patterns in parameter lists, but not *yet*. The parameter list grammar is already so complicated.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1374,173 +1580,172 @@ The static type of a pattern `p` being matched against a value of type `M` is:
+pattern variable declarations. The patterns that do type tests directly or
+implicitly can appear in variable declarations only if the tested type is a
+supertype of the value type. In other words, any pattern that needs to
+"downcast" to match is refutable.*
```
Ah, right. Done.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1557,7 +1762,7 @@ that contains the pattern:
+*   **Pattern-if statement**: The then statement.
 
 Multiple switch case patterns may share the same variable scope if their case
 bodies are empty:
```
While we're at it, let's delete the `<label>*` bit too. :)

Leaving it as-is for now.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1667,11 +1872,11 @@ behavior.
-2.  Match the `matcher` pattern against `v`.
+2.  Match the `pattern` against `v`.
 
 3.  If the match succeeds, evaluate the then `statement`. Otherwise, if there
```
I think that's covered in the scoping section.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1682,130 +1887,138 @@ At runtime, a pattern is matched against a value. This determines whether or not
+
+        3.  Otherwise, evaluate `v[k]` and match the resulting value against
+            this entry's value subpattern. If it does not match, the map does
+            not match.
```
In the previous step, we call `containsKey()` first to explicitly handle that. I think map patterns should distinguish between "absent key" and "key is mapped to `null`", so calling `containsKey()` explicitly does that.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -1816,9 +2029,9 @@ If a pattern variable declaration is marked `late` or a static variable
+pattern destructuring occurs and all uninitialized variables become initialized.
 
-*If you touch *any* of the variables, they *all* get initialized:*
+*If you read *any* of the variables, they *all* get initialized:*
```
Removed it.

munificent commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-listMatcher ::= ('<' typeOrBinder '>' )? '[' matchers ']'
+// This prints "Other order":
+switch (pair) {
+  case (a, b) when a > b:
```
I initially picked `if` for that exact reason but in practice, I do find `when` a *lot* easier to read.

Added a TODO to the proposal to iterate on this. Let's consider that a still-open issue to hash out separately.

munificent commented:
OK, I think I got through everything. That was an in-depth review. Thank you!

I revamped all of the static semantics stuff. I think it's a lot more coherent now.

Applied lots of small fixes you suggested. Removed support for stuff that you pointed out was dubious.

I added TODOs for the stuff that isn't resolved: null-check pattern syntax and guard syntax. The more general syntactic approach of not requiring `var` or `final` on binders is still up for discussion too, of course.

leafpetersen commented on working/0546-patterns/patterns-feature-specification.md:
```diff
@@ -144,953 +144,1059 @@ double calculateArea(Shape shape) =>
 
-**TODO: Should we support destructuring in `const` declarations?**
+A pair of patterns separated by `&` matches only if *both* subpatterns match.
+Unlike logical-or patterns, each branch may define different variables, since
```
may -> must?

munificent replied:
Reworded.

munificent commented:
I'm going to go ahead and merge this. I have other smaller unrelated changes I want to make to address open issues and I don't want to keep piling up more changes in this PR.

I think I've addressed all the minor things in your review comments, Lasse. Thank you! For the main unresolved stuff, there are TODOs in the proposal doc itself, and I still consider them unresolved. But we can merge and keep iterating forward. Even if we end up going back to something like the previous syntax where binders are marked explicitly with `var`, I still think this PR is a better starting point because it unifies the refutable and irrefutable grammars and fixes a bunch of stuff around static typing.
